### DISCUSSION TITLE
How to avoid infinite loop when extending a model which use the HasSettings trait

### DISCUSSION THREAD

**Discussion by felahdab** [glorand/laravel-model-settings#112] [Show and tell]
Hi everyone.

First: thank you to all developer and contributors to this package which is very well done and practical.

I want to share a problem into which I ran with this package. It doesn't really qualify as a bug. It's not about a feature request neither.

Long story short:
In a project I am working on, I have included this package and I am using the [nWidart/laravel-modules](https://github.com/nWidart/laravel-modules) package.

In order to separate concerns, I have models from different modules which extend models from other modules or from the base namespace (`App\Models`).

I faced the following situation:
- model `App\Models\User` uses `HasSettingsTable `(and therefore `HasSettings `trait also).
- model `Modules\HR\Entities\Person` extends the User module

When doing that without more precaution, I faced an infinite loop and therefore a sudden and silent application crash, when calling any method on the Person model resulting in a query having to be built and executed. 

[This infinite loop was coming from the `HasSettings` trait way of overriding the `__call` method in order to catch the `$invokeSettingsby` property when it is set. ](https://github.com/glorand/laravel-model-settings/blob/2fe95e4caa7bd7111d44cb4edbf219737ba8b409/src/Traits/HasSettings.php#L43C16-L43C83)

In particular, in that method, the use of `call_user_func(get_parent_class($this) . '::__call', $name, $args);` triggers the infinite loop.
I presume that since the Person model already extends the User model, but doesn't override the `__call` method which is already a method of `get_parent_class($this)`, the call_user_func is actually calling itself.

Even if it made itself visible when building a query from the inherited model, many other uses of that same model would have resulted in the infinite loop.

I am unsure as to why the `call_user_func `syntax was used in order to forward the calls to parent classes in `HasSettings`.
Changing this line to `parent::__call($name, $args)` seemed to work equally well (but I didn't go thru complete tests and might have missed something). 

But I didn't want to change the module code so I resorted to another way of resolving this issue without having to change the HasSettings code :
I simply override the __call method in the Person model like so:

```php
public function __call($name, $args)
{
        parent::__call($name, $args)
}
```
And by doing just that, since I make sure that the `HasSettings __call `method is called at the `User` level of inheritance, the infinite loop doesn't happen.

In conclusion:
- if anyone encounters a sudden silent application crash when using this module, it might have to do with an infinite loop caused by the supercharging of the `__call` method.
- maybe the syntax `parent::__call($name, $args)` would do equally well for the laravel-model-settings package, which I leave to the developpers to figure out.
- I hope that my trouble and this explanation will save some time for others.

Regards.
